helix-core/src/textobject.rs

```diff
@@ -45,6 +49,54 @@ fn find_word_boundary(slice: RopeSlice, mut pos: usize, direction: Direction) ->
     pos
 }
 
+fn find_paragraph_boundary(slice: RopeSlice, mut pos: usize, direction: Direction) -> usize {
+    // if pos is at non-empty line ending or when going forward move one character left
+    if (!char_is_line_ending(slice.char(pos.saturating_sub(1).min(slice.len_chars())))
+        || direction == Direction::Forward)
+        && char_is_line_ending(slice.char(pos.min(slice.len_chars().saturating_sub(1))))
+    {
+        pos = pos.saturating_sub(1);
+    }
+
+    let iter = chars_from_direction(&slice, pos, direction);
+
+    let prev_line_ending = match direction {
+        Direction::Forward => {
+            char_is_line_ending(slice.char(pos.saturating_sub(2)))
+                && char_is_line_ending(slice.char(pos.saturating_sub(1).min(slice.len_chars() - 1)))
+        }
+        Direction::Backward if pos == slice.len_chars() => true,
+        Direction::Backward => {
+            char_is_line_ending(slice.char(pos.saturating_sub(1)))
+                && char_is_line_ending(slice.char(pos))
+        }
+    };
+
+    // keep finding for two consecutive different line ending
+    // have to subtract later since we take past one or more cycle
+    let mut found = true;
+    let iter = iter.take_while(|&c| {
+        let now = prev_line_ending == char_is_line_ending(c);
+        let ret = found || now; // stops when both is different
+        found = now;
+        ret
+    });
+    let count = iter.count();
+    // count will be subtracted by extra whitespace due to interator
+    match direction {
+        Direction::Forward if pos + count == slice.len_chars() => slice.len_chars(),
+        // subtract by 1 due to extra \n when going forward
+        Direction::Forward if prev_line_ending => pos + count.saturating_sub(1),
+        Direction::Forward => pos + count,
+        // iterator exhausted so it should be 0
+        Direction::Backward if pos.saturating_sub(count) == 0 => 0,
+        // subtract by 2 because it starts with \n and have 2 extra \n when going backwards
+        Direction::Backward if prev_line_ending => pos.saturating_sub(count.saturating_sub(2)),
+        // subtract by 1 due to extra \n when going backward
+        Direction::Backward => pos.saturating_sub(count.saturating_sub(1)),
+    }
```

Comment:
@cessen Thansk for the review, after you remind me about CRLF I think this part have potential issues.

Reply:
Is this now resolved?

Reply:
This part is good, but the part above this haven't figure out how to switch it yet. https://github.com/helix-editor/helix/pull/784#discussion_r748699309

Also, I am still fixing on the goto next paragraph, haven't figured that out yet.